This CI failed with error message `"[2024-07-24T01:17:06Z] FAILED models/test_paligemma.py::test_models[5-128-float-size_factors0-google/paligemma-3b-mix-224] - TypeError: not a string"` It appears that this error is unrelated to the current PR,
I think this PR could be merged, cc @simon-mo

--- a/vllm/engine/llm_engine.py
+++ b/vllm/engine/llm_engine.py
@@ -949,8 +949,10 @@ class LLMEngine:
             model_output: Optional[List[SamplerOutput]] = None) -> None:
         """Forced log when no requests active."""
         if self.log_stats:
-            for logger in self.stat_loggers.values():
-                logger.log(self._get_stats(scheduler_outputs, model_output))
+            logger = self.stat_loggers["prometheus"]
+            logger.log(self._get_stats(scheduler_outputs, model_output))
+            logger = self.stat_loggers["logging"]
+            logger.log(self._get_stats(scheduler_outputs, model_output))
 
     def _get_stats(
             self,

--- a/vllm/engine/llm_engine.py
+++ b/vllm/engine/llm_engine.py
@@ -948,11 +948,9 @@ class LLMEngine:
             scheduler_outputs: Optional[SchedulerOutputs] = None,
             model_output: Optional[List[SamplerOutput]] = None) -> None:
         """Forced log when no requests active."""
-        if self.log_stats:
-            logger = self.stat_loggers["prometheus"]
-            logger.log(self._get_stats(scheduler_outputs, model_output))
-            logger = self.stat_loggers["logging"]
-            logger.log(self._get_stats(scheduler_outputs, model_output))
+        stats = self._get_stats(scheduler_outputs, model_output)
+        for logger in self.stat_loggers.values():
+            logger.log(stats)
 
     def _get_stats(
             self,

--- a/vllm/spec_decode/spec_decode_worker.py
+++ b/vllm/spec_decode/spec_decode_worker.py
@@ -484,7 +484,7 @@ class SpecDecodeWorker(LoraNotSupportedWorkerBase):
         for both speculation cases (num_lookahead_slots>0) and non-speculation
         cases (e.g. prefill).
 
-        Returns True iff there are remaining sequences to process.
+        Returns True if there are remaining sequences to process.
         """
         assert self.rank != self._driver_rank
 

